[DirectX] Remove trivially dead functions at linkage finalize

pow2clk opened the pull request:
Functions are not removed even when made internal by DXILFinalizeLinkage. The removal code is called from alwaysinliner and globalopt, which are invoked too early to remove functions made internal by this pass.

This adds a check similar to that in alwaysinliner that removes trivially dead functions after being marked internal. It refactors that code a bit to make it simpler including reversing what is stored int he work queue. 

Tests both the pass in isolation and the full i0nlining, linkage finalization and function removal steps. 

Fixes #106139

Commit:
[DirectX] Remove trivially dead functions at linkage finalize

Functions are not removed even when made internal by DXILFinalizeLinkage
The removal code is called from alwaysinliner and globalopt, which are
invoked too early to remove functions made internal by this pass.

This adds a check similar to that in alwaysinliner that removes
trivially dead functions after being marked internal. It refactors
that code a bit to make it simpler including reversing what is
stored in the work queue.

Tests both the pass in isolation and the full inlining, linkage
finalization, and function removal process.

Fixes #106139

pow2clk commented:
I'd like to add a test that verifies this removal of used functions that get alwaysinlined, but that requires the inlining fix for #89282

pow2clk commented on llvm/lib/Target/DirectX/DXILFinalizeLinkage.cpp:
```diff
@@ -18,19 +18,20 @@
+  for (Function *F : Funcs) {
+    if (F->getLinkage() == GlobalValue::ExternalLinkage) {
+      F->setLinkage(GlobalValue::InternalLinkage);
+      if (F->isDefTriviallyDead())
```
Not sure if this should also check that the alwaysinline attribute is set like AlwaysInliner does: https://github.com/llvm/llvm-project/blob/2a5ac9d9aff91406b0c58629df3a4e4dce87738c/llvm/lib/Transforms/IPO/AlwaysInliner.cpp#L89

We should always inline functions that are marked internal, so I don't know of a case where it wouldn't be set.

python3kgae commented on llvm/lib/Target/DirectX/DXILFinalizeLinkage.cpp:
```diff
@@ -19,20 +19,20 @@
+  for (Function *F : Funcs) {
+    if (F->getLinkage() == GlobalValue::ExternalLinkage)
+      F->setLinkage(GlobalValue::InternalLinkage);
+    if (F->hasFnAttribute(Attribute::AlwaysInline) && F->isDefTriviallyDead())
```
Could we remove F which is isDefTriviallyDead() but not AlwaysInline here?

pow2clk replied:
Yes probably. I backed away from doing that because removing this would require changing a lot of unrelated tests in llvm/test/CodeGen/DirectX so their funcions have some reason to stick around under the new rules. That would generally mean creating new calls to them or explicitly marking them noinline. Under ordinary circumstances, I expect alwaysinline to be set on all but the entry function anyway. It's not a very strong reason, but that's why I did it.

hekota replied:
Could these tests be updated to use `export` keyword to preserve the functions?

pow2clk replied:
Done!

Commit:
correct run lines

bob80905 commented on clang/test/CodeGenHLSL/remove-internal-unused.hlsl:
```diff
@@ -0,0 +1,47 @@
+     calledAndInlined();
+     calledAndNoinlined();
+     buf[0] = 0;
+}
```
new line at EOF

pow2clk replied:
Fixed

bob80905 commented on llvm/lib/Target/DirectX/DXILFinalizeLinkage.cpp:
```diff
@@ -19,20 +19,20 @@
-  SmallPtrSet<Function *, 8> EntriesAndExports;
+  SmallPtrSet<Function *, 8> Funcs;
 
   // Find all entry points and export functions
```
This comment may be better updated to something like:
Collect all non-entry point and non-exported functions to possibly set to internal linkage later.

pow2clk replied:
Updated

Commit:
Remove test requiring DirectX backend

hekota commented on llvm/test/CodeGen/DirectX/finalize-linkage-remove-dead.ll:
```diff
@@ -0,0 +1,80 @@
+}
+
+; Marked external, which should become internal and be removed.
+; CHECK: define internal void @"?doSomethingExternally@@YAXXZ"() #0
```
Same here

hekota commented on llvm/test/CodeGen/DirectX/finalize-linkage-remove-dead.ll:
```diff
@@ -0,0 +1,80 @@
+}
+
+; Marked internal, this should be removed.
+; CHECK: define internal void @"?doSomethingInternally@@YAXXZ"() #0
```
Typo - 'should not be removed'?

hekota commented on llvm/test/CodeGen/DirectX/finalize-linkage-remove-dead.ll:
```diff
@@ -0,0 +1,80 @@
+; is not explicitly marked noinline or export and is not an entry point.
+
+; Not called nor marked with any linking or inlining attributes.
+; CHECK-NOT: define {{.*}}doNothingNothing
```
Maybe add 'should be removed' to be consistent with all the other comments.

hekota commented on llvm/test/CodeGen/DirectX/finalize-linkage-remove-dead.ll:
```diff
@@ -0,0 +1,80 @@
+; Has noinline attribute. Should remain.
+; CHECK: define {{.*}}doNothingNoinline
+define void @"?doNothingNoinline@@YAXXZ"() #2 {
+entry:
```
Are there two (`doNothingDefault` and `doNothingNoinline`) not trivially dead when they are not called from anywhere?

pow2clk replied:
I've pretty dramatically changed this test in the most recent commit. I hope the comments are more understandable now.

Commit:
Eliminate trivially dead functions regardless of inline status

The consensus of the reviewers agreed with this approach

Adds better testing too

Commit:
Merge remote-tracking branch 'origin/main' into hlsl_remove_internal

farzonl commented on llvm/test/CodeGen/DirectX/scalar-load.ll:
```diff
@@ -48,11 +48,13 @@ define <4 x i32> @load_static_array_of_vec_test(i32 %index) {
   ret <4 x i32> %3
 }
+
+attributes #0 = { convergent norecurse nounwind "hlsl.export"}
```
Why are we add attributes we don't test?

pow2clk replied:
With this change, any functions lacking  this attribute will be removed because they are not called. The attribute is aded so they don't get removed and cause this and other tests to fail.

hekota commented on llvm/test/CodeGen/DirectX/finalize-linkage-remove-dead.ll:
```diff
@@ -0,0 +1,222 @@
+; RUN: opt -S -dxil-finalize-linkage -mtriple=dxil-unknown-shadermodel6.5-compute %s | FileCheck %s
+; RUN: llc %s --filetype=asm -o - | FileCheck %s
```
It might be good to have a similar test for libraries that would make sure all exported functions stay in the module despite not being called.

pow2clk replied:
Done!

hekota replied:
Shouldn't the exported functions stay in the module when it is compiled as a library, even though they are not called?

pow2clk replied:
Yes, and this test expects them to. The exported functions start at line 137 and all of them have checks that they remain defined.

Commit:
Add library DX codegen test

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_remove_internal

hekota reviewed (approved):
LGTM!

Commit:
Create fully separate library test

since compute shaders shouldn't have exported functions, we won't test them that way

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_remove_internal

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_remove_internal

Commit:
Correct mistaken merge
